File: app/src/main/java/com/youer/genshin/ui/MyFragment.java
package com.youer.genshin.ui;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.youer.genshin.App;
import com.youer.genshin.constants.Constants;
import com.youer.genshin.databinding.FragmentMyBinding;
import com.youer.genshin.resp.Result;
import com.youer.genshin.utils.SPUtil;
import com.youer.genshin.utils.StorageUtil;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * @author youer
 * @date 9/20/23
 */
public class MyFragment extends BaseFragment {

    FragmentMyBinding binding;
    private static final String TAG = "MyFragment";

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentMyBinding.inflate(getLayoutInflater());
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        binding.btCharacter.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                getCharacter();
            }
        });
        binding.btRelics.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                getRelics();
            }
        });
        binding.logout.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                SPUtil.removeKey(getContext(), Constants.KEY_UID);
                startActivity(new Intent(getContext(), LoginActivity.class));
                ((Activity)getContext()).finish();
            }
        });
        if (StorageUtil.fileExists(getContext(), Constants.FILE_CHARACTER)) {
            Constants.CHARACTERS = new Gson().fromJson(StorageUtil.readJsonFromFile(getContext(), Constants.FILE_CHARACTER), JsonObject.class);
        } else {
            getCharacter();
        }
        if (StorageUtil.fileExists(getContext(), Constants.FILE_RELICS)) {
            Constants.RELICS = new Gson().fromJson(StorageUtil.readJsonFromFile(getContext(), Constants.FILE_RELICS), JsonObject.class);
        } else {
            getCharacter();
        }
        getRelics();
    }

    /**
     * 加载角色list
     */
    public void getCharacter() {

        Call<Result<JsonObject>> call = App.getApiService().getCharacter();
        call.enqueue(new Callback<Result<JsonObject>>() {
            @Override
            public void onResponse(Call<Result<JsonObject>> call, Response<Result<JsonObject>> response) {
                if (response.isSuccessful()) {
                    Constants.CHARACTERS = response.body().getResult();
                    StorageUtil.saveJsonToFile(getContext(), new Gson().toJson(response.body().getResult()), Constants.FILE_CHARACTER);
                }
            }

            @Override
            public void onFailure(Call<Result<JsonObject>> call, Throwable t) {
                Log.e(TAG, "onFailure: ", t);
            }
        });
    }

    /**
     * 加载圣遗物list
     */
    public void getRelics() {

        Call<Result<JsonObject>> call = App.getApiService().getRelics();
        call.enqueue(new Callback<Result<JsonObject>>() {
            @Override
            public void onResponse(Call<Result<JsonObject>> call, Response<Result<JsonObject>> response) {
                if (response.isSuccessful()) {
                    Constants.RELICS = response.body().getResult();
                    StorageUtil.saveJsonToFile(getContext(), new Gson().toJson(response.body().getResult()), Constants.FILE_RELICS);
                }
            }

            @Override
            public void onFailure(Call<Result<JsonObject>> call, Throwable t) {
                Log.e(TAG, "onFailure: ", t);
            }
        });
    }
}

File: app/src/main/java/com/youer/genshin/ui/MyFragment.kt
package com.youer.genshin.ui

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.google.gson.Gson
import com.google.gson.JsonObject
import com.youer.genshin.App
import com.youer.genshin.constants.Constants
import com.youer.genshin.databinding.FragmentMyBinding
import com.youer.genshin.presenter.MainPresenter
import com.youer.genshin.resp.Result
import com.youer.genshin.utils.SPUtil
import com.youer.genshin.utils.StorageUtil
import org.json.JSONObject
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

/**
 * @author youer
 * @date 9/20/23
 */
class MyFragment(presenter: MainPresenter) : BaseFragment(presenter) {
    private lateinit var binding: FragmentMyBinding
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = FragmentMyBinding.inflate(layoutInflater)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.apply {
            btCharacter.setOnClickListener { character }
            btRelics.setOnClickListener { relics }
            logout.setOnClickListener {
                SPUtil.removeKey(context, Constants.KEY_UID)
                startActivity(Intent(context, LoginActivity::class.java))
                (context as Activity?)?.finish()
            }
        }
        buildCharacter();
    }

    fun buildCharacter(){
        Constants.CHARACTERS = JsonObject()
        for ((key, value) in Constants.CHARACTERS_INFO?.entrySet()!!) {
            // 在这里对每个键值对进行操作
            var tmp = (value as JsonObject).get("NameTextMapHash")
            if(tmp!=null) {
                Constants.CHARACTERS?.add(key, Constants.LOC_INFO?.get(tmp.asString))
            }
        }
    }

    /**
     * 加载角色list
     */
    private val character: Unit
        get() {
            val call: Call<Result<JsonObject>> = App.apiService.character
            call.enqueue(object : Callback<Result<JsonObject>> {
                override fun onResponse(call: Call<Result<JsonObject>>, response: Response<Result<JsonObject>>) {
                    if (response.isSuccessful) {
                        Constants.CHARACTERS = response.body().result
                        StorageUtil.saveJsonToFile(context, Gson().toJson(response.body().result), Constants.FILE_CHARACTER)
                        Toast.makeText(context, "更新成功", Toast.LENGTH_SHORT).show()
                    } else {
                        Toast.makeText(context, response.message(), Toast.LENGTH_SHORT).show()
                    }
                }

                override fun onFailure(call: Call<Result<JsonObject>>, t: Throwable) {
                    Log.e(TAG, "onFailure: ", t)
                }
            })
        }

    /**
     * 加载圣遗物list
     */
    private val relics: Unit
        get() {
            val call: Call<Result<JsonObject>> = App.apiService.relics
            call.enqueue(object : Callback<Result<JsonObject>> {
                override fun onResponse(call: Call<Result<JsonObject>>, response: Response<Result<JsonObject>>) {
                    if (response.isSuccessful) {
                        Constants.RELICS = response.body().result
                        StorageUtil.saveJsonToFile(context, Gson().toJson(response.body().result), Constants.FILE_RELICS)
                        Toast.makeText(context, "更新成功", Toast.LENGTH_SHORT).show()
                    } else {
                        Toast.makeText(context, response.message(), Toast.LENGTH_SHORT).show()
                    }
                }

                override fun onFailure(call: Call<Result<JsonObject>>, t: Throwable) {
                    Log.e(TAG, "onFailure: ", t)
                }
            })
        }

    companion object {
        private const val TAG = "MyFragment"
    }
}